**Logistic Regression**:

--> Response is not continuous
--> Response is categorical

3 responses:-

 **1. BINARY LOGISTIC REGRESSION**

--> when response is a binary data (Y/N)
 Binary Logistic Regression
-> predict when someone get job or not OR buy my insurance or not

**2. NOMINAL LOGISTIC REGRESSION**

--> response can have multiple categories
response order is not ordered.
atleast 3 options honge


**3. ORDINAL LOGISTIC REGRESSION**


--> This is used when the response has multiple categories and that categories have a logical order.

**LOGISTIC REGRESSION:**

**1)BINARY LOGISTIC REGRESSION**

Example for **binary logistic regression**:

In [2]:
import numpy as np
import pandas as pd
import statsmodels
import sklearn
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
import sklearn
from sklearn.model_selection import train_test_split


In [3]:
import os

In [4]:
os.chdir(r'/content/drive/MyDrive/STATSCDAC')

how the power transmission depend on hp,wt,am

ht,wt are predictors
am =response

In [5]:
df= pd.read_excel('CDAC_DataBook.xlsx', sheet_name='mtcars')

In [7]:
df.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [10]:
df = df[['hp','wt','am']]

In [11]:
df.head()

hp     wt  am
0  110  2.620   1
1  110  2.875   1
2   93  2.320   1
3  110  3.215   0
4  175  3.440   0

In [12]:
x_train = df.drop('am',axis=1)
y_train = df.am

In [13]:
x_train = sm.add_constant(x_train,prepend=False)

In [14]:
x_train.head()

hp     wt  const
0  110  2.620    1.0
1  110  2.875    1.0
2   93  2.320    1.0
3  110  3.215    1.0
4  175  3.440    1.0

In [20]:
mod1 = sm.Logit(y_train,x_train).fit()

Optimization terminated successfully.
         Current function value: 0.157174
         Iterations 9


In [19]:
print(mod1.summary())

                           Logit Regression Results                           
Dep. Variable:                     am   No. Observations:                   32
Model:                          Logit   Df Residuals:                       29
Method:                           MLE   Df Model:                            2
Date:                Sat, 24 Jun 2023   Pseudo R-squ.:                  0.7673
Time:                        09:59:05   Log-Likelihood:                -5.0296
converged:                       True   LL-Null:                       -21.615
Covariance Type:            nonrobust   LLR p-value:                 6.267e-08
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
hp             0.0363      0.018      2.044      0.041       0.001       0.071
wt            -8.0835      3.069     -2.634      0.008     -14.098      -2.069
const         18.8663      7.444      2.535      0.0

H0 for hp:
H0: am won't depend on hp

as p<0.05, H0 is rejected
hp influences am

H0 for wt:
H0: wt won't depend on hp

as p<0.05, H0 is rejected
wt influences am

coeff of wt represents as wt inc, am decreases.
as the coeff of wt increases, the prob of output tends to zero increases.



In [14]:
mydata  = pd.DataFrame([[120,2,1],[120,2.5,1],[120,3,1]],columns = ['hp','wt','const'])

In [15]:
mydata

hp   wt  const
0  120  2.0      1
1  120  2.5      1
2  120  3.0      1

In [16]:
mod1.predict(mydata)

0    0.999133
1    0.952947
2    0.262415
dtype: float64

here, the prob of output equal to 1 decreases. which means the prob of output equal to 0 increases.

**the coefficient of the wt represents the change in the log of odds when the wt increases by 1 ton.**

Another example:


In [21]:
df= pd.read_excel('CDAC_DataBook.xlsx', sheet_name='diabetes')

In [22]:
df = df[['Glucose','BloodPressure','Age','Outcome']]

In [23]:
x_train,x_test,y_train,y_test = train_test_split(df.drop('Outcome',axis=1),df['Outcome'],test_size=0.25)


In [24]:
x_train = sm.add_constant(x_train,prepend=False)

In [25]:
mod1 = sm.Logit(y_train,x_train).fit()

Optimization terminated successfully.
         Current function value: 0.506163
         Iterations 6


In [26]:
print(mod1.summary())

                           Logit Regression Results                           
Dep. Variable:                Outcome   No. Observations:                  576
Model:                          Logit   Df Residuals:                      572
Method:                           MLE   Df Model:                            3
Date:                Sat, 24 Jun 2023   Pseudo R-squ.:                  0.2120
Time:                        10:12:21   Log-Likelihood:                -291.55
converged:                       True   LL-Null:                       -370.00
Covariance Type:            nonrobust   LLR p-value:                 8.546e-34
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Glucose           0.0372      0.004      9.430      0.000       0.029       0.045
BloodPressure    -0.0035      0.005     -0.655      0.512      -0.014       0.007
Age               0.0314      0.009     

--> Glucose and age p<0.05 which means they are statistically significant.
they influence the diabetes.

In [27]:
x_test = sm.add_constant(x_test,prepend=False)

In [28]:
pred_y = mod1.predict(x_test)

In [29]:
pred_y[1:5]

#gives prob of 0 or 1

492    0.164218
313    0.215035
212    0.891026
185    0.896288
dtype: float64

In [30]:
y_test[1:5]

#gives the value whether 0 or 1

492    0
313    0
212    0
185    1
Name: Outcome, dtype: int64

In [31]:
res = []

for ctr in pred_y:
  if ctr<0:
    res.append(0)
  else:
    res.append(1)

res[:5]

[1, 1, 1, 1, 1]

In [33]:
from sklearn.metrics import confusion_matrix

In [37]:
confusion_matrix(y_test,res)

array([[  0, 117],
       [  0,  75]])

--> sir's confusion matrix has been exained in your notes! read your notes for better understaning.


In [35]:
from sklearn.metrics import classification_report

In [36]:
print(classification_report(y_test,res))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       117
           1       0.39      1.00      0.56        75

    accuracy                           0.39       192
   macro avg       0.20      0.50      0.28       192
weighted avg       0.15      0.39      0.22       192



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


sirs classification report has been explained in your notes

array([[  0, 117],
       [  0,  75]])

In [39]:
75/(0+75)

1.0

In [40]:
75/(75+0)

1.0

**NOMINAL LOGISTIC REGRESSION:**

--> **No of categories will be atleast three.**

SES:

1- lower class
2- middle class
3- upper class

prog:

1- voacational
2- general
3- academic

In [37]:
df= pd.read_excel('CDAC_DataBook.xlsx', sheet_name='nominal')

In [38]:
df.head()

ses  write  math  prog
0    1     35    41     1
1    2     33    41     2
2    3     39    44     3
3    1     37    42     1
4    2     31    40     2

In [39]:
df = df.drop('write',axis = 1)

In [40]:
df.head()

ses  math  prog
0    1    41     1
1    2    41     2
2    3    44     3
3    1    42     1
4    2    40     2

In [41]:
ses_dummy = pd.get_dummies(df['ses'],drop_first = True)

In [42]:
ses_dummy.head()

2  3
0  0  0
1  1  0
2  0  1
3  0  0
4  1  0

In [43]:
df = df.drop('ses',axis=1)

In [44]:
df = pd.concat([df,ses_dummy],axis=1)

# dummys are made for predictors.
# while making dummy for SES, SES=1 is choosen as reference.

In [45]:
df.head()

# SES=1 is taken as the reference event for making dummy for predictors.

math  prog  2  3
0    41     1  0  0
1    41     2  1  0
2    44     3  0  1
3    42     1  0  0
4    40     2  1  0

In [46]:
x_train,x_test,y_train,y_test = train_test_split(df.drop('prog',axis=1),df['prog'],test_size=0.25)

In [51]:
x_train = sm.add_constant(x_train,prepend=False)

In [52]:
mod1 = sm.MNLogit(y_train,x_train).fit()

#sm refers to the statsmodels library, which provides various statistical models and tools for data analysis.
#MNLogit is a function from the statsmodels library that creates a multinomial logistic regression model object.
#The MNLogit function is called with the arguments y_train and x_train, representing the dependent variable and the independent variables,
#    respectively.
#The multinomial logistic regression model is created but not yet fitted to the data. The model object mod1 is obtained, which represents
#    the multinomial logistic regression model.
#To fit the multinomial logistic regression model to the data, you need to call the fit() method on the mod1 object:

Optimization terminated successfully.
         Current function value: 0.681462
         Iterations 7


--> sm refers to the statsmodels library, which provides various statistical models and tools for data analysis.
MNLogit is a function from the statsmodels library that creates a multinomial logistic regression model object.
The MNLogit function is called with the arguments y_train and x_train, representing the dependent variable and the independent variables, respectively.
The multinomial logistic regression model is created but not yet fitted to the data. The model object mod1 is obtained, which represents the multinomial logistic regression model.
To fit the multinomial logistic regression model to the data, you need to call the fit() method on the mod1 object:

In [53]:
print(mod1.summary())

                          MNLogit Regression Results                          
Dep. Variable:                   prog   No. Observations:                  150
Model:                        MNLogit   Df Residuals:                      142
Method:                           MLE   Df Model:                            6
Date:                Wed, 21 Jun 2023   Pseudo R-squ.:                  0.3490
Time:                        06:05:57   Log-Likelihood:                -102.22
converged:                       True   LL-Null:                       -157.02
Covariance Type:            nonrobust   LLR p-value:                 2.463e-21
    prog=2       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
math          -0.0207      0.033     -0.630      0.529      -0.085       0.044
2              2.8267      0.633      4.469      0.000       1.587       4.066
3              2.4687      1.141      2.164      0.0

prog =1 (vocational) has been taken as reference event for interpretation because, in categorical data response will be given with respect to some reference event.


prog =2 how factors can influence choice between vocational and  general
**prog =2:**

2 number row in the summary indicate if you move fom ses=1 to ses=2, how will that movement effect the choice between prog 1(reference event)  and prog2(target event). the choice will be in terms of probability.

**H0 for math in prog2: **


**H0:** the math score does not have influence the choice between prog 1(reference event) and prog 2(target event).

math score does not influence the choice between prog 1 and 2 as p>0.05, H0 is not rejected.

**H0 for SES=2 and prog2:**
movement from SES = 1 and SES =2 does not have influence the choice between prog 1(reference event) and prog 2(target event).

SES 1 to SES 2 have influence the choice between prog 1 and 2 as p<0.05, H0 is rejected.



prog =3 how factors can influence choice between vocational and  academic

prog =3:
3 number row in the summary indicate if you move fom ses=1 to ses=3, how will that movement effect the choice between prog 1 and prog 3. the choice will be in terms of probability.

math marks:

**H0 for math in prog2:**

H0: the math score does not have influence the choice between prog 1(reference event) and prog 3(target event)
as the p<0.05, it has influence on prog =3.


as math score inceases, the prob of choosing academic increases, because prog=3 ==> academic.

H0 for SES=2 and prog3:**
movement from SES = 1 and SES =2 does not have influence the choice between prog 1(reference event) and prog 3(target event)

as p<0.05, it influences prog.

from the table:

**coefficient of SES=2 in prog 2:**

2.9567 = log(prob(prog 2)/prob(prog 1)) = log(prob(TARGET event)/prob(REFERENCE event))

prob(prog 2)/prob(prog 1)   > 1

so here prob(prog 2) > prob(prog 1)

this indicates when we go from ses1 to SES2 then, the prob of picking general course will be more than the prob of picking vocational course.

Predictor is the cause, response is the effect.

Categorical data response will always be with respect to some reference

In [54]:
x_test = sm.add_constant(x_test,prepend=False)

In [55]:
pred_y = mod1.predict(x_test)

In [57]:
pred_y[:5]

0         1         2
31   0.623591  0.351771  0.024639
191  0.010157  0.046626  0.943217
94   0.072865  0.785922  0.141213
150  0.042139  0.294401  0.663460
185  0.008420  0.037084  0.954497

In [58]:
x_train.head()

math  2  3  const
13     37  1  0    1.0
195    71  0  1    1.0
119    49  1  0    1.0
23     40  0  0    1.0
81     54  1  0    1.0

In [59]:
mydata = pd.DataFrame([[60,1,0,1],[70,1,0,1],[80,1,0,1]], columns=['math','2','3','const'])

In [60]:
mydata

math  2  3  const
0    60  1  0      1
1    70  1  0      1
2    80  1  0      1

In [61]:
mod1.predict(mydata)


#as the math marks are increasing from  ses 2(middle status) in prog2, the prob of choosing academic course is increasing.
# here  the column 0 = vocational, 1= general, 2=academic

0         1         2
0  0.047218  0.343816  0.608966
1  0.023871  0.141342  0.834787
2  0.009936  0.047842  0.942222

In [62]:
mydata = pd.DataFrame([[70,0,0,1],[70,1,0,1],[70,0,1,1]],columns=['math','2','3','const'])

#here math marks is constant,

In [63]:
mydata

# 0 0 means we're calculating for SES 0 low class
# 1 0 means we're calculating for SES 1 middle class
# 0 1 means we're calculating for SES 2 high class

math  2  3  const
0    70  0  0      1
1    70  1  0      1
2    70  0  1      1

In [64]:
mod1.predict(mydata)

#here row 0,1,2 represents SES low, medium, high classes
#column represents 0 = vocational, 1= general, 2=academic



0         1         2
0  0.573907  0.201201  0.224892
1  0.023871  0.141342  0.834787
2  0.006327  0.026190  0.967483

**VARIANCE INFLATION FACTOR(VIF):**
A variance inflation factor (VIF) is a measure of the amount of multicollinearity in regression analysis. Multicollinearity exists when there is a correlation between multiple independent variables in a multiple regression model. This can adversely affect the regression results. Thus, the variance inflation factor can estimate how much the variance of a regression coefficient is inflated due to multicollinearity..

Example

In [37]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='VIF')

In [11]:
df.head()

Education  Region  Gender  Exp  Union   Wage  Age  Race  Occupation  \
0          9       0       1   42      0   4.95   57     3           6   
1         12       0       0    1      0   6.67   19     3           6   
2         12       0       0    4      0   4.00   22     3           6   
3         12       0       0   17      0   7.50   35     3           6   
4         13       0       0    9      1  13.07   28     3           6   

   Sector  Married  
0       1        1  
1       1        0  
2       0        0  
3       0        1  
4       0        0

In [12]:
df.columns

Index(['Education', 'Region', 'Gender', 'Exp', 'Union', 'Wage', 'Age', 'Race',
       'Occupation', 'Sector', 'Married'],
      dtype='object')

In [38]:
df  = df[['Education','Gender','Exp','Age','Wage']]

In [21]:
gd_dummy = pd.get_dummies(df['Gender'],drop_first=True)

#gender is categorical data

In [39]:
df = df.drop('Gender',axis=1)

In [40]:
df = pd.concat([df,gd_dummy],axis=1)

In [42]:
df.head()

Education  Exp  Age   Wage  1
0          9   42   57   4.95  1
1         12    1   19   6.67  0
2         12    4   22   4.00  0
3         12   17   35   7.50  0
4         13    9   28  13.07  0

In [41]:
x_train = df.drop('Wage',axis=1)
x_train

Education  Exp  Age  1
0            9   42   57  1
1           12    1   19  0
2           12    4   22  0
3           12   17   35  0
4           13    9   28  0
..         ...  ...  ... ..
528         18    5   29  0
529         12   33   51  1
530         17   25   48  1
531         12   13   31  0
532         16   33   55  0

[533 rows x 4 columns]

In [43]:
y_train = df['Wage']
y_train

0       4.95
1       6.67
2       4.00
3       7.50
4      13.07
       ...  
528    11.36
529     6.10
530    23.25
531    19.88
532    15.38
Name: Wage, Length: 533, dtype: float64

In [44]:
mod1 = sm.OLS(y_train,x_train).fit()

In [45]:
print(mod1.summary())

                                 OLS Regression Results                                
Dep. Variable:                   Wage   R-squared (uncentered):                   0.817
Model:                            OLS   Adj. R-squared (uncentered):              0.816
Method:                 Least Squares   F-statistic:                              591.8
Date:                Wed, 21 Jun 2023   Prob (F-statistic):                   1.10e-193
Time:                        11:25:42   Log-Likelihood:                         -1550.9
No. Observations:                 533   AIC:                                      3110.
Df Residuals:                     529   BIC:                                      3127.
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

To check is there any variation influence factor is present or not?
import below module

In [31]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [32]:
variance_inflation_factor(df.values,0)

#col 0 is education

368.9046844691842

In [33]:
variance_inflation_factor(df.values,1)

#col 1 is experience

544.6987999131815

In [34]:
variance_inflation_factor(df.values,2)

#col 2 is age

1562.7050138396557

indeally speaking, if vif = 0 to 10, no action is required.
10-25, also we can ignore.

but above VIF >20 or 25,you cannot ignore.
in above case, inorder to handle high VIF, we will drop age column.

Now, the VIF of remaining columns will go down drastically.

In [46]:
x_train.head()

Education  Exp  Age  1
0          9   42   57  1
1         12    1   19  0
2         12    4   22  0
3         12   17   35  0
4         13    9   28  0

In [47]:
x_train = x_train.drop('Age',axis=1)

In [49]:
mod1 = sm.OLS(y_train,x_train).fit()

In [50]:
print(mod1.summary())

                                 OLS Regression Results                                
Dep. Variable:                   Wage   R-squared (uncentered):                   0.813
Model:                            OLS   Adj. R-squared (uncentered):              0.812
Method:                 Least Squares   F-statistic:                              768.2
Date:                Wed, 21 Jun 2023   Prob (F-statistic):                   1.73e-192
Time:                        11:26:54   Log-Likelihood:                         -1557.1
No. Observations:                 533   AIC:                                      3120.
Df Residuals:                     530   BIC:                                      3133.
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [51]:
variance_inflation_factor(x_train.values,0)
#col 1 is experience

2.9134749789780923

In [52]:
variance_inflation_factor(x_train.values,1)

#col 2 is age

2.4837216129295894

In [54]:
pip install mord

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for mord: filename=mord-0.7-py3-none-any.whl size=9885 sha256=2ca701610706a369a87d28eecdec34669bd757790f09aad41b1e88b71f130c7a
  Stored in directory: /root/.cache/pip/wheels/77/00/19/3cea86fbfc737ec4acb515cd94497dcc33f943fa157548b96c
Successfully built mord


In [55]:
import mord

**mord:**

--> "**Monotonic Ordinal Regression.**"
--> It is a package specifically designed for implementing ordinal regression models.

Ordinal regression is a type of regression analysis used when the dependent variable is ordinal, meaning it has ordered categories but the spacing or magnitude between categories may not be equal. The "mord" module provides various algorithms and methods to model and estimate ordinal regression models.

In [56]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='ordinal')

In [57]:
df.head()



Survival  Region  ToxicLevel
0         1       1        62.0
1         1       2        46.0
2         2       1        48.5
3         3       2        32.0
4         2       1        63.5

survival = 3 is good, 2 is better , 1is bad. there is logical order.

survival is dependent on region and toxiclevel.

In [58]:
from mord import LogisticAT

In [59]:
reg_dummy = pd.get_dummies(df['Region'],drop_first=True)

In [60]:
df = df.drop('Region',axis=1)

In [63]:
df = pd.concat([df,reg_dummy],axis=1)

In [64]:
df.head()

Survival  ToxicLevel  2
0         1        62.0  0
1         1        46.0  1
2         2        48.5  0
3         3        32.0  1
4         2        63.5  0

In [65]:
x_train,x_test,y_train,y_test = train_test_split(df.drop('Survival',axis=1),df['Survival'],test_size=0.25)

In [66]:
x_train = sm.add_constant(x_train,prepend=False)

In [67]:
mod1 = LogisticAT().fit(x_train,y_train)

In [68]:
x_test = sm.add_constant(x_test,prepend=False)

In [69]:
pred_y = mod1.predict(x_test)

In [71]:
from sklearn.metrics import confusion_matrix

In [72]:
print(confusion_matrix(y_test,pred_y))

[[ 0  6  0]
 [ 0 10  0]
 [ 0  3  0]]


-->** For this model, we don't get p values.**

Sir's confusion matrix is:

[[ 0  3  0]
 [ 4  9  2]
 [ 1  0  0]]

**COUNT MODELS:**

under this models, the data is discrete.

where we use these count models:

**case1**:
**where the output or response is in counts and that counts don't have much variation.**

then we use, **Poisson regression**



**1)POISSON REGRESSION:**

--> This model is used when the response is discrete with less variation

Example:

In [73]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='poisson')

In [74]:
df.head()

num_awards  prog  math
0           0     3    41
1           0     1    41
2           0     3    44
3           0     3    42
4           0     3    40

In [75]:
prog_dummy = pd.get_dummies(df['prog'],drop_first=True)

--> prog is categorical data, so we created dummies.

In [76]:
df= df.drop('prog',axis=1)

In [77]:
df = pd.concat([df,prog_dummy],axis=1)

In [78]:
df.head()

num_awards  math  2  3
0           0    41  0  1
1           0    41  0  0
2           0    44  0  1
3           0    42  0  1
4           0    40  0  1

In [79]:
x_train,x_test,y_train,y_test = train_test_split(df.drop('num_awards',axis=1),df['num_awards'],test_size=0.25)

In [80]:
x_train = sm.add_constant(x_train,prepend=False)

In [82]:
from statsmodels.discrete.discrete_model import Poisson as psn

here, x_train and y_train has to be clubbed together.

In [83]:
df_train = pd.concat([x_train,y_train],axis=1)

In [84]:
df_train.head()

#here 2 and 3 are prog2 and prog3

math  2  3  const  num_awards
129    56  1  0    1.0           1
133    57  1  0    1.0           0
17     38  0  1    1.0           0
144    61  1  0    1.0           1
164    61  1  0    1.0           1

In [85]:
df_train.columns

Index(['math', 2, 3, 'const', 'num_awards'], dtype='object')

In [86]:
df_train.rename(columns = {2:'prog2',3:'prog3'},inplace=True)

In [87]:
df_train.head()

math  prog2  prog3  const  num_awards
129    56      1      0    1.0           1
133    57      1      0    1.0           0
17     38      0      1    1.0           0
144    61      1      0    1.0           1
164    61      1      0    1.0           1

In [88]:
mod1 = psn.from_formula('num_awards ~ math + prog2 + prog3', data=df_train).fit()

Optimization terminated successfully.
         Current function value: 0.898370
         Iterations 6


In [89]:
print(mod1.summary())

                          Poisson Regression Results                          
Dep. Variable:             num_awards   No. Observations:                  150
Model:                        Poisson   Df Residuals:                      146
Method:                           MLE   Df Model:                            3
Date:                Wed, 21 Jun 2023   Pseudo R-squ.:                  0.2213
Time:                        12:24:29   Log-Likelihood:                -134.76
converged:                       True   LL-Null:                       -173.04
Covariance Type:            nonrobust   LLR p-value:                 1.663e-16
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -5.3758      0.744     -7.227      0.000      -6.834      -3.918
math           0.0735      0.012      6.097      0.000       0.050       0.097
prog2          1.0062      0.383      2.625      0.0

prog1 is taken as reference event,
prog2 and prog3  values are with reference to prog1


**H0 for math:**
--> the math score does not influence num_awards.

here p<0.05 ,H0 is rejected. hence, math marks influence the num awards(no of job offers).

--> for every unit increase in math marks, the job offers will increase by 0.0735

**H0 for prog2:**

--> the movement from prog1 to prog2 the no of job offers remain the same.
--> as p<0.05, **H0 is rejected**.
--> so when we move from prog1 to prog2 the job offers are changing.
when we move from prog1 to prog2, the job offers increase because coeff of prog2 is 1.0062.

**H0 for prog3:**
the movement from prog1 to prog3 the no of job offers remain the same.
as p>0.05, H0is not rejected.
so, if you join in prog1 or prog3 the job offers will remain same.

So, by this two cases, if you join prog2 the job offers will increase.